### PR TITLE
Template

### PR DESCRIPTION
**Pull request by SphericalPotatoInVacuum**

### CODE REVIEW

File: src/logger/simple.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+        self.loss_buff: dict[str, dict[str, list]] = defaultdict()
+        self.loss_buff['values'] = defaultdict(list)
+        self.loss_buff['sumlast'] = defaultdict(int)
+        self.loss_buff['sum'] = defaultdict(int)
+        self.metrics_buff: dict[str, dict[str, list]] = defaultdict()
+        self.metrics_buff['values'] = defaultdict(list)
+        self.metrics_buff['sumlast'] = defaultdict(int)
+        self.metrics_buff['sum'] = defaultdict(int)
+        self.tb_path: str = tb_path
+        self.train_iter = 1
```
Comment (E1eMenta) [inline src/logger/simple.py:23]: Use 0 instead of 1

File: src/logger/simple.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+        self.val_iter = 1
```
Comment (E1eMenta) [inline src/logger/simple.py:24]: Same

File: src/logger/simple.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+        self.wandb = wandb.init(project="TextDeepFake", entity="sphericalpotatoinvacuum")
```
Comment (E1eMenta) [inline src/logger/simple.py:25]: sphericalpotatoinvacuum?

File: src/logger/simple.py [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+
+    def log_train(self, losses: Optional[Dict[str, float]] = None, images: Optional[Dict[str, Tensor]] = None):
+        if self.train_iter == 1:
```
Comment (E1eMenta) [inline src/logger/simple.py:28]: 0 instead of 1

File: src/config/color.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+            test_dataset,
+            batch_size=1,
+            shuffle=False,
+            num_workers=8
+        )
+
+        point_batches = 500
+
+        logger = Logger(print_freq=2)
+        storage = Storage('./models/colorization')
```
[the remark this answers is not in the document]
Comment (grenlayk) [reply src/config/color.py:54]: Also change `models` to `checkpoints`

File: src/logger/simple.py [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+                self.loss_buff['sumlast'][loss_name] += loss_value
+
+        if self.train_iter % self.print_freq == 0:
+            self.end_time = time.time()
+            logger.info(f'Batch: {self.train_iter}')
+            logger.info(f'Processing time for last {self.print_freq} batches: {self.end_time - self.end_time:.3f}s')
+            for loss_name in self.loss_buff["sumlast"]:
+                logger.info(
+                    f'Average {loss_name} loss over last {self.print_freq} batches: {self.loss_buff["sumlast"][loss_name] / self.print_freq}')
+                self.wandb.log({f"{loss_name} loss": self.loss_buff["sumlast"][loss_name] / self.print_freq})
```
Comment (E1eMenta) [inline src/logger/simple.py:45]: Scalars should be logged each iteration

File: src/config/typeface.py [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+from torchvision.datasets import ImageFolder
+from torchvision.models.vgg import vgg19
+from torchvision import transforms as T
+
+
+class Config:
+    def __init__(self):
+        disk.login()
+        device = 'cuda:0' if torch.cuda.is_available() else 'cpu'
+        logger.info(f'Using device: {device}')
```
Comment (grenlayk) [inline src/config/typeface.py:22]: I don't think we should have it here -- we have Logger class

File: src/config/typeface.py [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+
+        model = vgg19(pretrained=True)
+        model.classifier[-1] = torch.nn.Linear(4096, 2500)
+        model = model.to(device)
+        criterion = CrossEntropyLoss()
+        optimizer = Adam(model.parameters(), lr=1e-4, weight_decay=5e-4)
+        storage = Storage('checkpoints/typeface')
+        metric_logger = Logger()
+
+        transform = T.Compose([
```
Comment (grenlayk) [inline src/config/typeface.py:36]: This should be in Dataset class, isn't it?

File: src/logger/simple.py [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+        if losses:
+            for loss_name, loss_value in losses.items():
+                self.loss_buff['values'][loss_name] += [loss_value]
+                self.loss_buff['sum'][loss_name] += loss_value
+                self.loss_buff['sumlast'][loss_name] += loss_value
+
+        if metrics:
+            for metric_name, metric_value in metrics.items():
+                self.metrics_buff['values'][metric_name] += [metric_value]
+                self.metrics_buff['sum'][metric_name] += metric_value
```
Comment (E1eMenta) [inline src/logger/simple.py:75]: Where do we use it?

File: src/config/typeface.py [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+            T.ToTensor(),
+            T.Resize((64, 196)),
+            T.Normalize(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225])
+        ])
+        train_dataset = ImageFolder('data/Typefaces/train', transform)
+        val_dataset = ImageFolder('data/Typefaces/val', transform)
```
Comment (grenlayk) [inline src/config/typeface.py:42]: `TypefaceDataset` class should be here

File: src/losses/simple.py [hunk 11]
```diff
@@ -0,0 +1,6 @@
+class Loss:
+    def __init__(self):
```
Comment (E1eMenta) [inline src/losses/simple.py:2]: Why do we need this class? It is identical to nn.Module.

File: src/logger/simple.py (hunk 3 above)
Comment (grenlayk) [inline src/logger/simple.py:25]: There will be a mess due to different users -- we need to agree on something common for all, to init it somehow from user's machine

File: src/models/simple.py [hunk 12]
```diff
@@ -0,0 +1,9 @@
+from torch import nn
```
Comment (E1eMenta) [inline src/models/simple.py:1]: Redundant

File: src/logger/simple.py [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+            self.start_time = time.time()
+            logger.info(
+                'Training started'
+            )
+        if losses:
+            for loss_name, loss_value in losses.items():
+                self.loss_buff['values'][loss_name] += [loss_value]
+                self.loss_buff['sumlast'][loss_name] += loss_value
+
+        if self.train_iter % self.print_freq == 0:
```
Comment (grenlayk) [inline src/logger/simple.py:38]: We log something once in `print_freq` batches but counter goes through all epochs. Maybe we need to clear it between epochs?

File: src/storage/simple.py [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+        '''
+        modules dict example:
+
+        {
+            'model': model,
+            'optimizer': optimizer,
+            'scheduler': scheduler
+        }
+        '''
+        self.epoch_count += 1
```
Comment (E1eMenta) [inline src/storage/simple.py:31]: Why do we need epoch counter?

File: src/logger/simple.py [hunk 15]
```diff
@@ -0,0 +1,121 @@
...
+            self.start_time = self.end_time
+            self.loss_buff['values'].clear()
+            self.loss_buff['sumlast'].clear()
+
+        if self.train_iter % self.image_freq == 0 and images:
+            # TensorBoard magic
+            pass
+
+        self.train_iter += 1
+
```
Comment (grenlayk) [inline src/logger/simple.py:56]: Maybe we should add just `info` method to log something that is not applicable to training or validation?

File: src/models/color.py [hunk 16]
```diff
@@ -0,0 +1,110 @@
+import torch
```
Comment (grenlayk) [inline src/models/color.py:1]: We should add source of this colorization, at least arxiv paper or something

File: README.md [hunk 17]
```diff
@@ -1,2 +1,154 @@
-# text-deep-fake
-Our implementation of [Text Style Brush](https://arxiv.org/pdf/2106.08385.pdf) architecture.
+# text-deep-fake design doc
```
Comment (grenlayk) [inline README.md:1]: Update version of readme, it is a little bit outdated

File: src/training/simple.py [hunk 18]
```diff
@@ -0,0 +1,64 @@
...
+        self.logger = logger
+        self.train_dataloader = train_dataloader
+        self.val_dataloader = val_dataloader
+        self.max_epoch = max_epoch
+
+    def train(self):
+        self.model.train()
+
+        for input, target in self.train_dataloder():
+            pred = self.model(input)
```
Comment (E1eMenta) [inline src/training/simple.py:36]: Where is zero_grad?

File: src/config/simple.py [hunk 19]
```diff
@@ -0,0 +1,24 @@
+from src.data.simple import SimpleDataset
```
Comment (grenlayk) [inline src/config/simple.py:1]: Consider moving "simple" files to separate branch only with these template files

File: src/training/simple.py [hunk 20]
```diff
@@ -0,0 +1,64 @@
...
+            loss.backward()
+            self.optimizer.step()
+            self.logger.log_train({'main': loss.item()}, {'input': input})
+
+        self.scheduler.step()
+
+    def validate(self, epoch):
+        self.model.eval()
+
+        losses = 0
```
Comment (E1eMenta) [inline src/training/simple.py:47]: redundunt

File: src/training/color.py [hunk 21]
```diff
@@ -0,0 +1,150 @@
...
+import cv2
+import numpy as np
+import torch
+import torchvision
+from loguru import logger
+from torchvision.utils import save_image
+
+
+class ColorizationTrainer:
+    def __init__(self,
```
Comment (grenlayk) [inline src/training/color.py:13]: Add typings

File: src/training/simple.py [hunk 22]
```diff
@@ -0,0 +1,64 @@
...
+        for input, target in self.train_dataloder():
+            pred = self.model(input)
+            loss = self.criterion(pred, target)
+            loss.backward()
+            self.optimizer.step()
+            self.logger.log_train({'main': loss.item()}, {'input': input})
+
+        self.scheduler.step()
+
+    def validate(self, epoch):
```
Comment (grenlayk) [inline src/training/simple.py:44]: Add typing

File: run.py [hunk 23]
```diff
@@ -0,0 +1,22 @@
...
+from pathlib import Path
+
+import click
+
+
+@click.command()
+@click.argument('config_file',
+                type=click.Path(exists=True, dir_okay=False, readable=True),
+                default='./src/config/color.py')
+def run(config_file):
```
Comment (grenlayk) [inline run.py:13]: Add typing

File: src/training/simple.py [hunk 24]
```diff
@@ -0,0 +1,64 @@
...
+
+        for input, target in self.val_dataloder():
+            pred = self.model(input)
+            loss = self.criterion(pred, target)
+            metric = metric(pred, target)
+            self.logger.log_val({'loss':loss}, {'metric':metric}, {'input': input})
+
+        avg_losses, avg_metrics = self.logger.end_val()
+
+        self.storage.save(epoch, {}, avg_metrics['metric'])
```
Comment (E1eMenta) [inline src/training/simple.py:58]: Why the modules parameter is empty?

File: src/data/color.py [hunk 25]
```diff
@@ -0,0 +1,49 @@
...
+class CustomDataset(Dataset):
+    def __init__(self, root_dir: Path):
+        self.root_dir = root_dir
+        self.files = list(root_dir.iterdir())
+        logger.info(f'File[0]: {self.files[0]}, Total Files: {len(self.files)}')
+
+    def __len__(self):
+        return len(self.files)
+
+    def __getitem__(self, index):
```
Comment (grenlayk) [inline src/data/color.py:20]: Add typing

File: src/disk/disk.py [hunk 26]
```diff
@@ -0,0 +1,83 @@
...
+            config['YaDiskCreds'] = {
+                'secret': secret,
+                'token': token
+            }
+            config.write(self._cred_cache_path.open('w', encoding='utf-8'))
+
+        self._y = y
+        self._logged_in = True
+
+        logger.info('Logged in to YandexDisk')
```
Comment (grenlayk) [inline src/disk/disk.py:53]: Why don't we use Logger from outside?

File: src/training/simple.py (hunk 18 above)
Comment (E1eMenta) [inline src/training/simple.py:36]: Move input and target to deivce

File: src/training/typeface.py [hunk 27]
```diff
@@ -0,0 +1,77 @@
+from typing import Callable
+
+import torch
+from src.logger.simple import Logger
+from src.storage.simple import Storage
+from torch import nn, optim
+from torch.utils.data import DataLoader
+
+
```
Comment (E1eMenta) [inline src/training/typeface.py:9]: We can use simple.py

File: src/storage/simple.py [hunk 28]
```diff
@@ -0,0 +1,38 @@
...
+        if isinstance(save_folder, str):
+            self.save_path = Path(save_folder)
+        else:
+            self.save_path = save_folder
+
+        self.save_path.mkdir(parents=True, exist_ok=True)
+        self.save_freq = save_freq
+        self.epoch_count = 0
+
+    def save(self, epoch: int, modules: Dict[str, nn.Module], metric: float):
```
Comment (grenlayk) [inline src/storage/simple.py:21]: `metric` is unused

File: src/training/color.py [hunk 29]
```diff
@@ -0,0 +1,150 @@
...
+            # y: ab_img
+            for idx, (l_img, ab_img) in enumerate(self.train_dataloader):
+                # Skip bad data
+                if not l_img.ndim:
+                    continue
+
+                l_img = l_img.to(self.device)
+                ab_img = ab_img.to(self.device)
+
+                # Initialize Optimizer
```
Comment (grenlayk) [inline src/training/color.py:66]: So many comments -- do we need them all?

File: src/training/color.py [hunk 30]
```diff
@@ -0,0 +1,150 @@
...
+
+            # Adding l channel to ab channels
+            color_img = self.concatenate_and_colorize(torch.stack([l_img[:, 0, :, :]], dim=1), output_ab)
+
+            save_path = Path('outputs')
+            save_path.mkdir(exist_ok=True)
+            save_path /= f'img{idx}.jpg'
+            save_image(color_img[0], save_path)
+
+            # Printing to Tensor Board
```
Comment (grenlayk) [inline src/training/color.py:141]: Delete commented code

File: src/config/typeface.py [hunk 31]
```diff
@@ -0,0 +1,60 @@
...
+        self.trainer = TypefaceTrainer(
+            model=model,
+            criterion=criterion,
+            optimizer=optimizer,
+            scheduler=None,
+            storage=storage,
+            logger=metric_logger,
+            train_dataloader=train_dataloader,
+            val_dataloader=val_dataloader,
+            max_epoch=5,
```
Comment (E1eMenta) [inline src/config/typeface.py:55]: 5 epoch is a too small

File: src/training/typeface.py [hunk 32]
```diff
@@ -0,0 +1,77 @@
...
+            pred = self.model(inputs)
+            loss = self.criterion(pred, target)
+            loss.backward()
+            self.optimizer.step()
+            self.logger.log_train({'CE': loss.item()}, {'input': inputs})
+
+        if self.scheduler is not None:
+            self.scheduler.step()
+
+    def validate(self, epoch):
```
Comment (grenlayk) [inline src/training/typeface.py:52]: Add typing

File: src/logger/simple.py (hunk 3 above)
Comment (grenlayk) [reply src/logger/simple.py:25]: project -- as a param in config, entity -- team name from Wanda

File: src/config/typeface.py [hunk 33]
```diff
@@ -0,0 +1,60 @@
...
+from src.disk import disk
+from src.logger.simple import Logger
+from src.storage.simple import Storage
+from src.training.typeface import TypefaceTrainer
+from torch.nn import CrossEntropyLoss
+from torch.optim import Adam
+from torch.utils.data import DataLoader
+from torchvision.datasets import ImageFolder
+from torchvision.models.vgg import vgg19
+from torchvision import transforms as T
```
[the remark this answers is not in the document]
Comment (grenlayk) [reply src/config/typeface.py:15]: Seems to be common name -- let it be so then.

File: src/disk/disk.py (hunk 26 above)
Comment (grenlayk) [reply src/disk/disk.py:53]: Because they are different, okay

File: src/logger/simple.py [hunk 34]
```diff
@@ -0,0 +1,121 @@
...
+                    f'Average {loss_name} over validation: {self.loss_buff["sum"][loss_name] / (self.val_iter - 1)}')
+                avg_losses[loss_name] = self.loss_buff["sum"][loss_name] / (self.val_iter - 1)
+        if self.metrics_buff:
+            for metric_name in self.metrics_buff['sum']:
+                logger.info(
+                    f'Average {metric_name} over validation: {self.metrics_buff["sum"][metric_name] / (self.val_iter - 1)}')
+                avg_metrics[metric_name] = self.metrics_buff["sum"][metric_name] / (self.val_iter - 1)
+
+        for dict_name in self.loss_buff:
+            self.loss_buff[dict_name].clear
```
Comment (grenlayk) [inline src/logger/simple.py:114]: clear()

File: src/logger/simple.py (hunk 9 above)
Comment (grenlayk) [reply src/logger/simple.py:75]: in end val

File: src/data/typeface.py [hunk 35]
```diff
@@ -0,0 +1,48 @@
+import os
```
Comment (grenlayk) [inline src/data/typeface.py:1]: We don't need it
